packages/funding-rate-proposer/src/proposer.js

```diff
@@ -0,0 +1,360 @@
+    // We can execute updates for each perpetual contract in parallel because they
+    // should not affect each other.
+    await Promise.map(Object.keys(this.contractCache), contractAddress => {
+      return this._updateFundingRate(contractAddress);
+    });
+  }
+
+  /** **********************************
+   *
+   * INTERNAL METHODS
+   *
+   ************************************/
+
+  // Check contract funding rates and request+propose to update them, or return early if an error is encountered.
+  async _updateFundingRate(contractAddress) {
+    const cachedContract = this.contractCache[contractAddress];
+    const currentFundingRateData = cachedContract.state.currentFundingRateData;
+    const currentConfig = cachedContract.state.currentConfig;
+    const fundingRateIdentifier = this.hexToUtf8(currentFundingRateData.identifier);
+
+    // If proposal time is not 0, then proposal is already outstanding. Check if
+    // the proposal has been disputed and if not, then we can't propose and must exit.
+    const proposalTime = currentFundingRateData.proposalTime.toString();
+    if (proposalTime !== "0") {
+      this.logger.debug({
+        at: "PerpetualProposer#updateFundingRate",
+        message: "Proposal is already pending, cannot propose",
+        fundingRateIdentifier,
+        proposalTime
+      });
+      return;
+    }
+
+    // Assume pricefeed has been cached and updated prior to this function via the `update()` call.
+    const priceFeed = this.priceFeedCache[fundingRateIdentifier];
+    let offchainFundingRate = priceFeed.getCurrentPrice().toString();
+    if (!offchainFundingRate) {
+      this.logger.error({
+        at: "PerpetualProposer#updateFundingRate",
+        message: "Failed to query current price for funding rate identifier",
+        fundingRateIdentifier
+      });
+      return;
+    }
+    let onchainFundingRate = currentFundingRateData.rate.toString();
+
+    // Check that offchainFundingRate is within [configStore.minFundingRate, configStore.maxFundingRate]
+    const minFundingRate = currentConfig.minFundingRate.toString();
+    const maxFundingRate = currentConfig.maxFundingRate.toString();
+    if (
+      this.toBN(offchainFundingRate).lt(this.toBN(minFundingRate)) ||
+      this.toBN(offchainFundingRate).gt(this.toBN(maxFundingRate))
+    ) {
+      this.logger.error({
+        at: "PerpetualProposer#updateFundingRate",
+        message: "Potential proposed funding rate is outside allowed funding rate range",
+        fundingRateIdentifier,
+        minFundingRate,
+        maxFundingRate,
+        offchainFundingRate
+      });
+      return;
+    }
+
+    // If the saved funding rate is not equal to the current funding rate within margin of error, then
+    // prepare request to update. We're assuming that the `offchainFundingRate` is the baseline or "expected"
+    // price.
+    let isPriceDisputable = isDeviationOutsideErrorMargin(
```

Comment:
isPriceDisputable seems like the wrong name for this variable...? shoulduUdateFundingRate, maybe?

Reply:
Bump.